custom/ewsghana/reports/specific_reports/stock_status_report.py

```diff
@@ -382,6 +339,156 @@ class StockStatus(MultiReport):
+
+        if not supply_points:
+            return {}
+
+        unique_products = self.unique_products(locations, all=True)
+        transactions = StockTransaction.objects.filter(
+            type='stockonhand',
+            case_id__in=list(supply_points),
+            report__domain=self.report_config['domain'],
+            report__date__lte=self.report_config['enddate'],
+            report__date__gte=self.report_config['startdate'],
+            product_id__in=list(unique_products.values_list('product_id', flat=True)),
+        ).distinct('case_id', 'product_id').order_by('case_id', 'product_id', '-report__date').values_list(
+            'case_id', 'product_id', 'report__date', 'stock_on_hand'
+        )
+
+        stockout_table_supply_points = list(self.get_supply_points.values_list('supply_point_id', flat=True))
+        grouped_by_case = defaultdict(set)
+        stock_states = StockState.objects.filter(
+            sql_product__domain=self.domain,
+            case_id__in=stockout_table_supply_points
+        )
+        product_case_with_stock = defaultdict(set)
+        product_case_without_stock = defaultdict(set)
+
+        months_of_stock = defaultdict(lambda: defaultdict(dict))
+        stock_state_map = {
+            (stock_state.case_id, stock_state.product_id):
+            stock_state.get_monthly_consumption() if stock_state.daily_consumption else None
+            for stock_state in stock_states
+        }
+
+        current_mos_locations = set(self.get_supply_points.values_list('supply_point_id', flat=True))
+        stockouts = defaultdict(set)
+        for (case_id, product_id, date, stock_on_hand) in transactions:
+            if stock_on_hand > 0:
+                product_case_with_stock[product_id].add(case_id)
+                if case_id in current_mos_locations:
+                    stock_state_dict = stock_state_map.get((case_id, product_id))
+                    if stock_state_dict:
+                        months_of_stock[case_id][product_id] = stock_on_hand / stock_state_dict
+                    else:
+                        months_of_stock[case_id][product_id] = None
+            else:
+                product_case_without_stock[product_id].add(case_id)
+                if case_id in current_mos_locations:
+                    stockouts[case_id].add(product_id)
+                    months_of_stock[case_id][product_id] = 0
+            grouped_by_case[case_id].add(product_id)
+
+        for case_id in current_mos_locations:
+            if case_id not in months_of_stock:
+                for product in unique_products:
+                    consumption = stock_state_map.get((case_id, product.product_id))
+                    try:
+                        stock_on_hand = StockTransaction.objects.filter(
+                            case_id=case_id,
+                            product_id=product.product_id
+                        ).latest('report__date').stock_on_hand
```

Comment:
@esoergel You are not right about complexity of nesting for loops. This for loop has complexity O(m * n) where m is number of products and n is number of locations. Worst scenario is m = 41 on national level. 

It's hard to say how many times it's called because this loop is executed only for locations that didn't report stock in particular period